components/browser/state/src/main/java/mozilla/components/browser/state/reducer/WebExtensionReducer.kt

```diff
@@ -0,0 +1,136 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.state.reducer
+
+import mozilla.components.browser.state.action.WebExtensionAction
+import mozilla.components.browser.state.selector.findTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.SessionState
+import mozilla.components.browser.state.state.TabSessionState
+import mozilla.components.browser.state.state.WebExtensionState
+
+internal object WebExtensionReducer {
+    /**
+     * [WebExtensionAction] Reducer function for modifying a specific [WebExtensionState] in
+     *  both [SessionState] or [BrowserState].
+     */
+    fun reduce(state: BrowserState, action: WebExtensionAction): BrowserState {
+        return when (action) {
+            is WebExtensionAction.TabState -> {
+                ExtensionTabStateReducer.reduce(state, action)
+            }
+            is WebExtensionAction.BrowserState -> {
+                ExtensionBrowserStateReducer.reduce(state, action)
+            }
+        }
+    }
+}
+
+private object ExtensionBrowserStateReducer {
+    /**
+     * [WebExtensionAction] Reducer function for modifying a specific [WebExtensionState] in
+     * [SessionState]
+     */
+    fun reduce(state: BrowserState, action: WebExtensionAction.BrowserState): BrowserState {
+        return when (action) {
+            is WebExtensionAction.BrowserState.InstallWebExtension -> {
+                // Verify that the new extension doesn't already exist
+                requireUniqueExtension(state.extensions, action.extension)
+                state.copy(
+                    extensions = state.extensions + action.extension
+                )
+            }
+            is WebExtensionAction.BrowserState.UpdateBrowserAction -> {
+                state.updateExtensionState(action.extensionId) {
+                    it.copy(browserAction = action.browserAction)
+                }
+            }
+        }
+    }
+
+    private fun BrowserState.updateExtensionState(
+        extensionId: String,
+        update: (WebExtensionState) -> WebExtensionState
+    ): BrowserState {
+        return copy(
+            extensions = extensions.map { current ->
+                if (current.id == extensionId) {
+                    update.invoke(current)
+                } else {
+                    current
+                }
+            }
+        )
+    }
+}
+
+private object ExtensionTabStateReducer {
+    /**
+     * [WebExtensionAction] Reducer function for modifying a specific [WebExtensionState] in
+     * [BrowserState]
+     */
+    fun reduce(state: BrowserState, action: WebExtensionAction.TabState): BrowserState {
+        return when (action) {
+            is WebExtensionAction.TabState.InstallWebExtension -> {
```

Comment:
As discussed, we don't need this action. When we get a session-specific override, we can add it to the list if it's not already there.